consensus/src/consensus_observer/pending_blocks.rs

```diff
@@ -3,100 +3,202 @@
+};
 use std::{collections::BTreeMap, sync::Arc};
 
 /// A simple struct to store the block payloads of ordered and committed blocks
 #[derive(Clone)]
 pub struct PendingOrderedBlocks {
-    // Pending ordered blocks (indexed by consensus round)
-    pending_ordered_blocks: Arc<Mutex<BTreeMap<Round, (OrderedBlock, Option<CommitDecision>)>>>,
+    // The configuration of the consensus observer
+    consensus_observer_config: ConsensusObserverConfig,
+
+    // Verified and unverified pending ordered blocks. The key is the epoch and
+    // round of the last block in the ordered block. Each entry contains the
+    // block, if the block was verified, and the commit decision (if any).
+    pending_blocks:
+        Arc<Mutex<BTreeMap<(u64, Round), (OrderedBlock, bool, Option<CommitDecision>)>>>,
 }
 
 impl PendingOrderedBlocks {
-    pub fn new() -> Self {
+    pub fn new(consensus_observer_config: ConsensusObserverConfig) -> Self {
         Self {
-            pending_ordered_blocks: Arc::new(Mutex::new(BTreeMap::new())),
+            consensus_observer_config,
+            pending_blocks: Arc::new(Mutex::new(BTreeMap::new())),
         }
     }
 
-    /// Clears all pending blocks
-    pub fn clear_all_pending_blocks(&self) {
-        self.pending_ordered_blocks.lock().clear();
-    }
-
-    /// Returns a copy of the pending ordered blocks map
-    pub fn get_all_pending_blocks(
+    /// Returns a copy of the verified pending blocks
+    pub fn get_all_verified_pending_blocks(
         &self,
-    ) -> BTreeMap<Round, (OrderedBlock, Option<CommitDecision>)> {
-        self.pending_ordered_blocks.lock().clone()
+    ) -> BTreeMap<(u64, Round), (OrderedBlock, Option<CommitDecision>)> {
+        let mut verified_pending_blocks = BTreeMap::new();
+        for (key, (ordered_block, verified_ordered_proof, commit_decision)) in
+            self.pending_blocks.lock().iter()
+        {
+            if *verified_ordered_proof {
+                verified_pending_blocks
+                    .insert(*key, (ordered_block.clone(), commit_decision.clone()));
+            }
+        }
+        verified_pending_blocks
     }
 
-    /// Returns the pending ordered block (if any)
+    /// Returns the last pending ordered block (if any). We take into
+    /// account verified and unverified pending blocks (to ensure we're
+    /// able to buffer blocks across epoch boundaries).
     pub fn get_last_pending_block(&self) -> Option<BlockInfo> {
-        let pending_ordered_blocks = self.pending_ordered_blocks.lock();
-        if let Some((_, (ordered_block, _))) = pending_ordered_blocks.last_key_value() {
-            Some(ordered_block.blocks.last().unwrap().block_info())
-        } else {
-            None // No pending blocks were found
-        }
+        self.pending_blocks
+            .lock()
+            .last_key_value()
+            .map(|(_, (ordered_block, _, _))| ordered_block.last_block().unwrap().block_info())
```

Comment:
Use `expect` instead of `unwrap`? (#13992)

Reply:
Be aware `expect` works exactly like `unwrap`, **panic**, and just provide a message in the form of answering “what I expect?”. A proper error handling or default value is the best choice!

Reply:
We now `expect` in the `first_block()` and `last_block()` methods 😄

Reply:
Unfortunately, avoiding the use of "expect" in my PR proved to be very challenging. It would have required significant changes throughout the codebase to ensure proper error propagation. The decision to ban "unwrap" but allow "expect" was made to strike a balance; it provides a tool for initiating reflection (and documenting it) without being overly restrictive. If you would like to discuss this further, I would be happy to continue the conversation in my PR.